libraries/chain_kv/include/b1/session/key_value.hpp

```diff
@@ -0,0 +1,148 @@
+// \returns A key_value instance.
+// \remarks This factory does not guarentee that the the memory used for the key/value will be contiguous in memory.
+inline key_value make_kv(bytes key, bytes value) {
+    auto result = key_value{};
+    result.m_key = std::move(key);
+    result.m_value = std::move(value);
+    return result;
+}
+
+// Instantiaties a key_value instance from the given key/value pointers.
+//
+// \tparam key The pointer type of the key.
+// \tparam value The pointer type of the value.
+// \tparam allocator The memory allocator type used to manage the bytes memory. This type mush implement the "memory allocator" concept.
+// \param the_key A pointer to the key.
+// \param key_length The size of the memory pointed by the key.
+// \param the_value A pointer to the value.
+// \param value_length The size of the memory pointed by the value.
+// \param a The memory allocator used to managed the memory used by the key_value instance.
+// \returns A key_value instance.
+// \remarks This factory guarentees that the memory needed for the key and value will be contiguous in memory.
+template <typename key, typename value, typename allocator>
+key_value make_kv(const key* the_key, size_t key_length, const value* the_value, size_t value_length, allocator& a) {
+    auto total_key_length = key_length * sizeof(key);
+    auto total_value_length = value_length * sizeof(value);
+    return make_kv(reinterpret_cast<const void*>(the_key), total_key_length, reinterpret_cast<const void*>(the_value), total_value_length, a);
+}
+
+template <typename key, typename value>
+key_value make_kv(const key* the_key, size_t key_length, const value* the_value, size_t value_length) {
+    return make_kv(reinterpret_cast<const void*>(the_key), key_length * sizeof(key), reinterpret_cast<const void*>(the_value), value_length * sizeof(key));
+}
+
+inline key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length) {
+    auto kv = key_value{};
+    kv.m_key = make_bytes(key, key_length);
+    kv.m_value = make_bytes(value, value_length);
+    return kv;
+}
+
+
+// Instantiaties a key_value instance from the given key/value pointers.
+//
+// \tparam allocator The memory allocator type used to manage the bytes memory. This type mush implement the "memory allocator" concept.
+// \param key A pointer to the key.
+// \param key_length The size of the memory pointed by the key.
+// \param value A pointer to the value.
+// \param value_length The size of the memory pointed by the value.
+// \param a The memory allocator used to managed the memory used by the key_value instance.
+// \returns A key_value instance.
+// \remarks This factory guarentees that the memory needed for the key and value will be contiguous in memory.
+template <typename allocator>
+key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length, allocator& a) {
+    // TODO:  Not quite sure how to manage this memory chunk.
```

Comment:
To make the key and the value contiguous, you can allocate a single buffer and set its initial reference count to 2, letting it be shared by `m_key` and `m_value`.

Reply:
Perfect!